### ISSUE TITLE
Logging errors

### ISSUE THREAD

**Issue by kohearn-cwf** [Canadian-Wildlife-Federation/CHYF#20]
Hi Emily, I just wanted to create a new issue for this (following up on our email conversation) since we had a team call today to discuss any other issues that had come up regarding logs. I have a few more examples of AOIs you can use for testing if needed, listed below. Unless otherwise noted, all the log files are available on the automatic processing server.

### AOI fails without logging errors or updating status in DB:
* 0701000 (flowpathgenerator-20221109-123002-0701000.log)
* 07LC000 (flowpathgenerator-20221109-113002-07LC000.log)
* 10CD000 (flowpathgenerator-20221108-143002-10CD000.log)

### Processing time not writing to the logs
* 10JE000 ([flowpathgenerator-20221108-135022-10JE000.log](https://github.com/Canadian-Wildlife-Federation/CHYF/files/9974364/flowpathgenerator-20221108-135022-10JE000.log))
* 07LA000 (flowpathgenerator-20221108-143002-07LA000.log)

### Errors not writing back to the DB:
* 10JE000 ([flowpathgenerator-20221108-135022-10JE000.log](https://github.com/Canadian-Wildlife-Federation/CHYF/files/9974364/flowpathgenerator-20221108-135022-10JE000.log)) - Log file included many errors and warnings, but none of these were written back to the DB
* 10JA001 ([flowpathgenerator-20221108-135124-10JA001.log](https://github.com/Canadian-Wildlife-Federation/CHYF/files/9974386/flowpathgenerator-20221108-135124-10JA001.log)) - Error for "Cannot add path as both directions for this path creates a cycle." did not get written back to the database but all other errors/warnings were present. I am testing a theory with this workunit today to see if the errors I saw in the DB were simply from an old run.
* 10MB000 ([flowpathgenerator-20221108-153802-10MB000.log](https://github.com/Canadian-Wildlife-Federation/CHYF/files/9974410/flowpathgenerator-20221108-153802-10MB000.log)) - This workunit and 1001000 encountered the same error, I suspect this is similar to the Java heap space error where the tools are entirely stopped from running?
* 1001000 ([flowpathgenerator-20221108-153802-1001000.log](https://github.com/Canadian-Wildlife-Federation/CHYF/files/9974412/flowpathgenerator-20221108-153802-1001000.log))

**Comment by kohearn-cwf** [comment]
Follow up on testing for 10JA001 where error was not getting written back to the database. Cleared all errors from fpoutput.errors for this AOI before running the flowpath tools again. All errors except for "Cannot add path as both directions for this path creates a cycle" were present in the DB, I am not sure why this particular error is not being written back. The AOI still gets a status of FP_ERROR in the fpoutput.aoi table.

**Comment by egouge** [comment]
@kohearn-cwf for 10MB000  and 1001000  (Errors not writing back to the DB).   These look to be some issue creating the local geopackage file used for processing. A few things to check out
1) Both of these have at least one null geometry in the catchments table. I'm guessing this might be the problem:

select count(*) from fpinput.ecatchment where aoi_id ='b1d080eb-d794-4857-8ae0-327cdc3307d5'  and geometry is null;
select count(*) from fpinput.ecatchment where aoi_id ='d08ea9fd-2be0-4622-a29d-976b06255bff'  and geometry is null;

If resolving this fixes the error, we should add this as a QA check to the QA scripts.

2) If that doesn't help - at the top of the log files you'll see this message: "Creating geopackage file:C:\Users\kohearn\AppData\Local\Temp\fp12710499599815410086.geopkg".  For each run the geopackage file is named a a bit different so you'll need to run it then look at the log file to get the file for that run. The question is does the file exists and if you look at the contents in QGIS do the contents look correct (it should be a copy of the data in the fpinput schema for the aoi that's being processed)?

**Comment by kohearn-cwf** [comment]
> ### AOI fails without logging errors or updating status in DB:
> * 0701000 (flowpathgenerator-20221109-123002-0701000.log)
> * 07LC000 (flowpathgenerator-20221109-113002-07LC000.log)
> * 10CD000 (flowpathgenerator-20221108-143002-10CD000.log)

Hi Emily, I re-ran 07LC000 this morning using the v.1.5.6 of the tools on automatic processing server to see if this issue was resolved. Processing stopped with the following line:
> 2022-11-14 15:12:57.761 [pool-1-thread-2] INFO  n.r.c.f.s.voronoi.SkeletonJob - ECatchments.7387

The AOI status remains as FP_PROCESSING and no errors were logged. I suspect this could be a heap space error, although I have not seen that error appear for this workunit when running it locally. I will leave this workunit for testing and move on to another for now!

**Comment by egouge** [comment]
@kohearn-cwf regarding the feature_names_aoi_id_unq error - Is is possible that two different processing tools were processing this aoi at the same time? This is the only reason I can think of that this error should occur.  This table is cleared before processing starts so the only way the values can be duplicated is if somebody else wrote to that table before the processing finished.

**Comment by kohearn-cwf** [comment]
> @kohearn-cwf regarding the feature_names_aoi_id_unq error - Is is possible that two different processing tools were processing this aoi at the same time? This is the only reason I can think of that this error should occur. This table is cleared before processing starts so the only way the values can be duplicated is if somebody else wrote to that table before the processing finished.

It shouldn't be possible, as we have only been running the tools using the status as the indicator for processing (i.e., we don't use the -a flag at all). 

I noticed that the name_id referenced in the error message was present in the rivernameid2 field for several skeleton flowpaths in the original NHN data that shouldn't have had that value assigned. I have removed those values, so my latest run should hopefully indicate if that was the source of the issue. If the error does show up again, I'll check with the team once this latest run is finished to see if anyone was somehow able to pick it up on their run as well. Agreed that this is a really strange error.
